services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/events.py

```diff
@@ -503,114 +506,133 @@ async def action(cls, app: FastAPI, scheduler_data: SchedulerData) -> None:
         # invoke container cleanup at this point
         dynamic_sidecar_client: DynamicSidecarClient = get_dynamic_sidecar_client(app)
 
-        try:
-            await dynamic_sidecar_client.begin_service_destruction(
-                dynamic_sidecar_endpoint=scheduler_data.dynamic_sidecar.endpoint
-            )
-        except BaseClientHTTPError as e:
-            logger.warning(
-                (
-                    "Could not contact dynamic-sidecar to begin destruction of "
-                    "%s\n%s. Will continue service removal!"
-                ),
-                scheduler_data.service_name,
-                f"{e}",
-            )
-        app_settings: AppSettings = app.state.settings
-        dynamic_sidecar_settings: DynamicSidecarSettings = (
-            app_settings.DYNAMIC_SERVICES.DYNAMIC_SIDECAR
-        )
-
-        # only try to save the status if :
-        # - the dynamic-sidecar has finished booting correctly
-        # - it is requested to save the state
-        if (
-            scheduler_data.dynamic_sidecar.service_removal_state.can_save
-            and await are_all_services_present(
-                node_uuid=scheduler_data.node_uuid,
-                dynamic_sidecar_settings=app.state.settings.DYNAMIC_SERVICES.DYNAMIC_SIDECAR,
+        # A node can end up with all the services from a single study.
+        # When the study is closed, all the services will try to save their
+        # data. This causes a lot of disk and network stress.
+        # Some nodes collapse under load or behave unexpectedly.
```

Comment:
Q: what about the download of these data on start? is this not triggering the same kind of issues?

Reply:
Q2: did you find a way to simulate it and test/measures the iowait?

Reply:
Before it gets lost let me post the test here.

```python
# pylint: disable=redefined-outer-name
from typing import Iterator, Iterable
from pathlib import Path
from uuid import uuid4
import psutil
import time
from contextlib import contextmanager
from collections import deque
from threading import Thread
import tempfile
import json
from multiprocessing import Process, Queue

import pytest
from pydantic import ByteSize, parse_obj_as
from shutil import copyfileobj, COPY_BUFSIZE
import os

from requests import patch

# UTILS


def _get_random_file_name() -> str:
    return f"file_{uuid4()}"


def _gen_chunks(size: int, *, preferred_chunk_size=8 * 1024) -> Iterable[int]:
    full_chunks = size // preferred_chunk_size
    for _ in range(full_chunks):
        yield preferred_chunk_size
    remainder = size % preferred_chunk_size
    if remainder > 0:
        yield remainder


def _generate_random_file(target_dir: Path, size: ByteSize) -> Path:
    file_name = target_dir / _get_random_file_name()
    print(f"generating file {file_name}")
    with open(file_name, "wb") as output_file:
        for chunk in _gen_chunks(size):
            output_file.write(os.urandom(chunk))
    print(f"generated {file_name}")
    return file_name


@contextmanager
def get_aiowait_profile(interval: float) -> Iterator[deque[float]]:
    results: deque[float] = deque()
    can_continue: bool = True
    # run something her in parallel?
    def _worker() -> None:
        while can_continue:
            times = psutil.cpu_times_percent()
            results.append(times.iowait)
            time.sleep(interval)

    thread = Thread(target=_worker, daemon=True)
    # start profiling
    thread.start()
    try:
        yield results
    finally:
        # stop profiling
        can_continue = False
        thread.join()


class Counter:
    def __init__(self) -> None:
        self.start = time.time()
        self.end = 0.0

    def finished(self):
        self.end = time.time()

    @property
    def elapsed(self) -> str:
        elapsed = self.end - self.start
        return "%.02f" % elapsed


@contextmanager
def duration() -> Iterator[Counter]:
    c = Counter()
    yield c
    c.finished()


# FIXTURES


@pytest.fixture
def content_dir() -> Iterable[Path]:
    with tempfile.TemporaryDirectory() as temp_dir:
        yield Path(temp_dir)


@pytest.fixture
def big_file(content_dir: Path) -> Path:
    file_name = _generate_random_file(
        target_dir=content_dir, size=parse_obj_as(ByteSize, "1gib")
    )
    return Path(file_name)


@pytest.fixture(scope="session")
def results_file(here: Path) -> Path:
    path = here / "benchmark.ignore.json"
    if path.exists():
        path.unlink()
    return path


# TESTS


def test_get_aiowait_profile() -> None:
    with get_aiowait_profile(0.1) as something:
        time.sleep(1)
    assert len(something) > 0
    assert sum(something) == 0


@pytest.mark.parametrize("total_size", ["1b", "1025b", "1kib", "1mib"])
@pytest.mark.parametrize("preferred_chunk", ["12b", "1024b", "123b", "1gib"])
def test_gen_chunks(total_size: str, preferred_chunk: str) -> None:
    size = parse_obj_as(ByteSize, total_size)
    chunk_size = parse_obj_as(ByteSize, preferred_chunk)
    assert sum(x for x in _gen_chunks(size, preferred_chunk_size=chunk_size)) == size


def _copy_file_worker(
    source_path: Path,
    file_size: int,
    chunk_size: int,
    results_queue: Queue,
    profile_interval: float = 0.1,
) -> None:
    with tempfile.TemporaryDirectory() as temp_dir:
        target_path = Path(temp_dir) / _get_random_file_name()

        with get_aiowait_profile(profile_interval) as profile:
            with open(source_path, "rb") as source, open(
                target_path, "wb"
            ) as destination:

                with duration() as counter:
                    # NOTE: this is the actual function being benchmarked!
                    # copyfileobj(source, destination, parse_obj_as(ByteSize, chunk_size))
                    command = f"cp {source_path} {target_path}"
                    print(command, flush=True)
                    os.system(command)

        result = dict(
            duration=counter.elapsed,
            chunk_size=chunk_size,
            file_size=file_size,
            profile_interval=profile_interval,
            profile=list(profile),
        )
        results_queue.put(result)


@pytest.mark.parametrize("chunk_size", ["8kib", "64kib", "1024kib", "16mib"])
@pytest.mark.parametrize("file_size", ["1gib", "2gib"])
@pytest.mark.parametrize("parallel_copies", [1, 10])
def test_profile_cases(
    content_dir: Path,
    file_size: str,
    chunk_size: int,
    parallel_copies: int,
    results_file: Path,
) -> None:
    file_to_copy = _generate_random_file(
        target_dir=content_dir,
        size=parse_obj_as(ByteSize, file_size),
    )

    results_queue = Queue()
    processes = []
    for _ in range(parallel_copies):
        process = Process(
            target=_copy_file_worker,
            args=(file_to_copy, file_size, chunk_size, results_queue),
            daemon=True,
        )
        process.start()
        processes.append(process)

    for p in processes:
        p.join()

    assert results_queue.qsize() == parallel_copies

    data_in_file = []
    if results_file.exists():
        data_in_file = json.loads(results_file.read_text() or "[]")

    test_data = [results_queue.get() for _ in range(results_queue.qsize())]
    all_data = data_in_file + test_data
    print(len(all_data))
    results_file.write_text(json.dumps(all_data))


# TODO: plot this
# TODO: use a different implementation and see if it is better!
# TODO: also try this with CP!

# links:
# https://github.com/SickChill/SickChill/issues/3292
# https://stackoverflow.com/questions/22078621/how-to-copy-files-fast/28129677#28129677
```